[Game] Improves physics engine with Jitter2 upgrade and introduce lazy loading for terrain

ZeromusXYZ opened the pull request:
Upgrades the physics engine to Jitter2 for improved simulation accuracy and performance. Also added a lazy loading option for terrain.

Key changes include:
- Migrates from Jitter.Core to Jitter2, leveraging its performance enhancements.
- Introduces cell-based heightmap management for lazy terrain loading, reducing initial world load times and memory usage.
- Implements a new PhysicsManager to handle physics simulation within world instances.
- Updates ship physics to use ShipController for managing forces and movement.
- Addresses AIManager ticking to avoid null reference exceptions.
- Adds option to disable preloading terrain, improving startup time.

ZeromusXYZ committed:
[Game] Added lazy terrain loading

- Added option to allow loading heightmap data on-demand only be disabling pre-loading the entire world.

ZeromusXYZ committed:
[Game] Split up heightmap data into per cell data

- Removed the global ushort array for the world into smaller per cell arrays. This allows better lazy loading of the heightmaps.

ZeromusXYZ committed:
[Game] Upgrade to Jitter2

Co-authored-by: NLObP

- Moved to JitterPhysics2 for the boats
- Added on-demand loading for the physics engine
- Cleaned some code and added more comments

ZeromusXYZ committed:
[Game] WIP Physics

- Experimental changes to physics

ZeromusXYZ committed:
[Game] Better shudown

- Added checks to AI where it will be less likely to act upon non-existing Owners.
- Added despawns when shutting down
- Added a `AIManager.Instance.Stop()` to the shutdown process.

ZeromusXYZ committed:
[Game] Tweaks to boat physics

- Changed code so the Physics can handle proper delta times and changed it so TPS can be dynamic.
- Acceleration and deceleration feel better now.

Note:
- Turn speed has not been modified/updated yet and is way too fast.
- Reversing is still applying immediate breaking.

ZeromusXYZ committed:
[Game] More tweaks to boat physics

- More adjustments to speed and turn speed

ZeromusXYZ committed:
[Game] Add Configuration for Physics TPS

- Added options to `world.json` to set a `TargetPhysicsTps` value (25 by default)
- Added more timing info when spawning to get a better idea of how long it takes for the server to start.

ZeromusXYZ committed:
[Test] Updated code so the test can run

mgpreston commented on AAEmu.Game/Core/Managers/World/PhysicsManager.cs:
```diff
@@ -0,0 +1,604 @@
+
+                    // 3. Step the physics world
+                    // Potentially step multiple times to catch up if we were running behind.
+                    _physWorld.Step((float)physicsTotalDelta.TotalSeconds, false);
```
This has lost some of the changes from https://github.com/AAEmu/AAEmu/pull/1205/files
According to [the docs](https://jitterphysics.com/docs/documentation/world#time-step-size), the physics world should be stepped at a fixed delta, potentially multiple times, and not at a variable delta

mgpreston commented on AAEmu.Game/Configurations/World.json:
```diff
@@ -13,6 +13,8 @@
-        "MaxInstances": 32
+        "PreLoadTerrain": false,
+        "MaxInstances": 32,
+        "TargetPhysicsTps": 25.0
```
[The docs](https://jitterphysics.com/docs/documentation/world#time-step-size) say "Timesteps larger then dt=1/60s are not adviced", so this should be at least 60 TPS according to Jitter, but maybe no big deal if it's working well enough at 25

ZeromusXYZ replied:
I think it's more in the aspect if you actual have a GUI where you'd normally target at 60 fps.
We only send generic position and velocity updates to the game-client and that one does it's own calculations.
Also why I find that the "variable" tick-rate is working so well. (it's only variable if it wasn't able to complete the previous loop in time)

ZeromusXYZ committed:
[Game] Tweaking more values

- Fixes the actual move speed of ships to be more correct. It's not the correct way to fix it, but it'll have to do.

angelof9 commented:
The steering is kinda weird and seems to be linked to travel speed.

Changing steering direction seems to be inverse proportional to travel speed.
While continuous steering the same direction seems to increase with travel speed.

In short, to change steering direction while already turning you need to slow down the boat or it takes ages to change from left to right and vice/versa.

ZeromusXYZ committed:
[Game] Rebased onto develop and Jitter2 version update

ZeromusXYZ committed:
[Game] Fixed rivers

- Fixed the code where it wasn't doing checks anymore for anything other than the ocean.
- Restored flow direction (although not entirely correct)

ZeromusXYZ committed:
[Game] Tweaked the code for handling river flows

Last one for this PR, I promise 😅
